File: mobile/library/swift/stats/TagsBuilder.swift
import Foundation

/// Builder class used to construct `Tags` instances.
/// See `TagsBuilder` for usage.
@objcMembers
public class TagsBuilder: NSObject {
  private(set) var tags: [String: String]

  /// Append a value to the tag name.
  ///
  /// - parameter name:  The tag name.
  /// - parameter value: The value associated to the tag name.
  ///
  /// - returns: This builder.
  @discardableResult
  public func add(name: String, value: String) -> Self {
    self.tags[name] = value
    return self
  }

  /// Replace all values at the provided name with a new set of tag values.
  ///
  /// - parameter name: The tag name.
  /// - parameter value: The value associated to the tag name.
  ///
  /// - returns: This builder.
  @discardableResult
  public func set(name: String, value: String) -> Self {
    self.tags[name] = value
    return self
  }

  /// Remove all tags with this name.
  ///
  /// - parameter name: The tag name to remove.
  ///
  /// - returns: This builder.
  @discardableResult
  public func remove(name: String) -> Self {
    self.tags[name] = nil
    return self
  }

  /// Add all tags from dictionary to builder.
  ///
  /// - parameter name: a dictionary of tags.
  ///
  /// - returns: This builder.
  @discardableResult
  public func putAll(tags: [String: String]) -> Self {
    self.tags = self.tags.merging(tags, uniquingKeysWith: { _, last in last })
    return self
  }

  /// Build the tags using the current builder.
  ///
  /// - returns: New instance of request headers.
  public func build() -> Tags {
    return Tags(tags: self.tags)
  }

  // MARK: - Internal

  /// Allows for setting tags that are not publicly mutable (i.e., restricted tags).
  ///
  /// - parameter name: The tag name.
  /// - parameter value: The value associated to the tag name.
  ///
  /// - returns: This builder.
  @discardableResult
  func internalSet(name: String, value: String) -> Self {
    self.tags[name] = value
    return self
  }

  // Only explicitly implemented to work around a swiftinterface issue in Swift 5.1. This can be
  // removed once envoy is only built with Swift 5.2+
  override required init() {
    self.tags = [String: String]()
    super.init()
  }

  /// Initialize a new builder. Subclasses should provide their own public convenience initializers.
  ///
  /// - parameter tags: The tags with which to start.
  required init(tags: [String: String]) {
    self.tags = tags
    super.init()
  }
}

// MARK: - Equatable

extension TagsBuilder {
  public override func isEqual(_ object: Any?) -> Bool {
    return (object as? Self)?.tags == self.tags
  }
}

File: mobile/library/kotlin/io/envoyproxy/envoymobile/stats/TagsBuilder.kt
package io.envoyproxy.envoymobile

/*
 * Builder class used to construct `Tags` instances.
 */
public class TagsBuilder {
  protected val tags: MutableMap<String, String>

  /**
   * Instantiate a new builder.
   *
   * @param Tags: The Tags to start with.
   */
  public constructor(tags: MutableMap<String, String>) {
    this.tags = tags
  }

  public constructor() {
    this.tags = mutableMapOf<String, String>()
  }

  /**
   * Append a value to the Tag name.
   *
   * @param name: The Tag name.
   * @param value: The value associated to the Tag name.
   * @return TagsBuilder, This builder.
   */
  public fun add(name: String, value: String): TagsBuilder {
    tags.getOrPut(name) { value }
    return this
  }

  /**
   * Replace all values at the provided name with a new set of Tag values.
   *
   * @param name: The Tag name.
   * @param value: The value associated to the Tag name.
   * @return TagsBuilder, This builder.
   */
  public fun set(name: String, value: String): TagsBuilder {
    tags[name] = value
    return this
  }

  /**
   * Remove all Tags with this name.
   *
   * @param name: The Tag name to remove.
   * @return TagsBuilder, This builder.
   */
  public fun remove(name: String): TagsBuilder {
    tags.remove(name)
    return this
  }

  /**
   * Adds all tags from map to this builder.
   *
   * @param tags: A map of tags.
   * @return TagsBuilder, This builder.
   */
  public fun putAll(tags: Map<String, String>): TagsBuilder {
    this.tags.putAll(tags)
    return this
  }

  /**
   * Build the tags using the current builder.
   *
   * @return Tags, New instance of Tags.
   */
  public fun build(): Tags {
    return Tags(tags)
  }
}
